File: macos/Gutenkern/Sources/GutenkernCore/GlyphClassifier.swift
import Foundation

public struct ParsedToken: Equatable, Sendable {
    public var glyph: Glyph
    public var start: Int
    public var length: Int
}

public struct ClassificationResult: Equatable, Sendable {
    public var groups: [KerningGroup]
    public var glyphs: [KerningGroup: [Glyph]]
    public var unknown: [ParsedToken]

    public func glyphs(for group: KerningGroup) -> [Glyph]? {
        glyphs[group]
    }
}

public enum GlyphClassifier {
    private enum Kind: Equatable {
        case unknown
        case capital
        case lowercase
        case smallCaps
        case punctuation
        case nonAlphabetic
        case liningPlain
        case liningPnum
        case oldstyleOnum
        case oldstyleProportional
    }

    public static func classify(_ input: String) -> ClassificationResult {
        let tokens = KerningGenerator.parseTokens(input)
        var unknown: [ParsedToken] = []
        var tagged: [(Glyph, Kind)] = []

        for token in tokens {
            let kind = classifyToken(token.glyph)
            if kind == .unknown {
                unknown.append(token)
                continue
            }
            tagged.append((token.glyph, kind))
        }

        let hasPnum = tagged.contains { $0.1 == .liningPnum }
        let hasProportionalOldstyle = tagged.contains { $0.1 == .oldstyleProportional }
        var buckets: [KerningGroup: [Glyph]] = [:]
        for group in KerningGroup.allCases {
            buckets[group] = []
        }

        for (glyph, kind) in tagged {
            let group: KerningGroup?
            switch kind {
            case .capital: group = .capitals
            case .lowercase: group = .lowercase
            case .smallCaps: group = .smallCaps
            case .punctuation: group = .punctuation
            case .nonAlphabetic: group = .nonAlphabetic
            case .liningPlain: group = hasPnum ? nil : .liningFigures
            case .liningPnum: group = .liningFigures
            case .oldstyleOnum: group = hasProportionalOldstyle ? nil : .oldstyleFigures
            case .oldstyleProportional: group = .oldstyleFigures
            case .unknown: group = nil
            }
            if let group {
                buckets[group, default: []].append(glyph)
            }
        }

        let present = KerningGroup.allCases.filter { !(buckets[$0] ?? []).isEmpty }
        let glyphs = Dictionary(uniqueKeysWithValues: present.map { ($0, buckets[$0] ?? []) })
        return ClassificationResult(groups: present, glyphs: glyphs, unknown: unknown)
    }

    private static func classifyToken(_ glyph: Glyph) -> Kind {
        switch glyph {
        case .character(let character):
            let value = String(character)
            if let classified = GlyphDictionary.class(byChar: value) {
                return fromDictionary(classified)
            }
            return GlyphDictionary.isAsciiDigit(value) ? .liningPlain : .unknown
        case .name(let name):
            return classifyName(name)
        }
    }

    private static func classifyName(_ name: String) -> Kind {
        let parts = name.split(separator: ".", omittingEmptySubsequences: true).map(String.init)
        guard !parts.isEmpty else {
            return .unknown
        }

        let suffixes = Set(parts.dropFirst())
        if !suffixes.isDisjoint(with: ["sc", "smcp", "c2sc"]) {
            return .smallCaps
        }

        let baseName = parts[0]
        let hasPnum = !suffixes.isDisjoint(with: ["pnum", "lf"])
        let hasOnum = !suffixes.isDisjoint(with: ["onum", "tosf"])
        let hasOsf = suffixes.contains("osf")
        let hasTf = suffixes.contains("tf")
        if GlyphDictionary.isDigitName(baseName) && (hasPnum || hasOnum || hasOsf || hasTf) {
            if hasOsf || (hasPnum && hasOnum) {
                return .oldstyleProportional
            }
            if hasOnum {
                return .oldstyleOnum
            }
            if hasPnum {
                return .liningPnum
            }
            return .liningPlain
        }

        if let exact = GlyphDictionary.class(byName: name) {
            return fromDictionary(exact)
        }
        if let byBase = GlyphDictionary.class(byName: baseName) {
            return fromDictionary(byBase)
        }
        return .unknown
    }

    private static func fromDictionary(_ classification: String) -> Kind {
        switch classification {
        case "capital": return .capital
        case "lowercase": return .lowercase
        case "digit": return .liningPlain
        case "punctuation": return .punctuation
        case "nonalphabetic": return .nonAlphabetic
        default: return .unknown
        }
    }
}

File: windows/Gutenkern.Core/GlyphClassifier.cs
namespace Gutenkern;

public readonly record struct ParsedToken(Glyph Glyph, int Start, int Length);

public sealed class ClassificationResult
{
    public IReadOnlyList<KerningGroup> Groups { get; }
    public IReadOnlyDictionary<KerningGroup, IReadOnlyList<Glyph>> Glyphs { get; }
    public IReadOnlyList<ParsedToken> Unknown { get; }

    internal ClassificationResult(
        IReadOnlyList<KerningGroup> groups,
        IReadOnlyDictionary<KerningGroup, IReadOnlyList<Glyph>> glyphs,
        IReadOnlyList<ParsedToken> unknown)
    {
        Groups = groups;
        Glyphs = glyphs;
        Unknown = unknown;
    }

    public bool TryGet(KerningGroup group, out IReadOnlyList<Glyph> glyphs) =>
        Glyphs.TryGetValue(group, out glyphs!);
}

public static class GlyphClassifier
{
    private enum Kind
    {
        Unknown,
        Capital,
        Lowercase,
        SmallCaps,
        Punctuation,
        NonAlphabetic,
        LiningPlain,
        LiningPnum,
        OldstyleOnum,
        OldstyleProportional
    }

    public static ClassificationResult Classify(string input)
    {
        var tokens = KerningGenerator.ParseTokens(input);
        var unknown = new List<ParsedToken>();
        var tagged = new List<(Glyph Glyph, Kind Kind)>();

        foreach (var token in tokens)
        {
            var kind = ClassifyToken(token.Glyph);
            if (kind == Kind.Unknown)
            {
                unknown.Add(token);
                continue;
            }

            tagged.Add((token.Glyph, kind));
        }

        var hasPnum = tagged.Any(item => item.Kind == Kind.LiningPnum);
        var hasProportionalOldstyle = tagged.Any(item => item.Kind == Kind.OldstyleProportional);
        var buckets = Enum.GetValues<KerningGroup>().ToDictionary(
            group => group,
            _ => new List<Glyph>());

        foreach (var (glyph, kind) in tagged)
        {
            var group = kind switch
            {
                Kind.Capital => KerningGroup.Capitals,
                Kind.Lowercase => KerningGroup.Lowercase,
                Kind.SmallCaps => KerningGroup.SmallCaps,
                Kind.Punctuation => KerningGroup.Punctuation,
                Kind.NonAlphabetic => KerningGroup.NonAlphabetic,
                Kind.LiningPlain when hasPnum => (KerningGroup?)null,
                Kind.LiningPlain => KerningGroup.LiningFigures,
                Kind.LiningPnum => KerningGroup.LiningFigures,
                Kind.OldstyleOnum when hasProportionalOldstyle => (KerningGroup?)null,
                Kind.OldstyleOnum => KerningGroup.OldstyleFigures,
                Kind.OldstyleProportional => KerningGroup.OldstyleFigures,
                _ => null
            };
            if (group is { } resolved)
            {
                buckets[resolved].Add(glyph);
            }
        }

        var present = Enum.GetValues<KerningGroup>()
            .Where(group => buckets[group].Count > 0)
            .ToList();
        var glyphs = present.ToDictionary(
            group => group,
            group => (IReadOnlyList<Glyph>)buckets[group]);
        return new ClassificationResult(present, glyphs, unknown);
    }

    private static Kind ClassifyToken(Glyph glyph)
    {
        switch (glyph)
        {
            case CharacterGlyph character:
                if (GlyphDictionary.TryClassByChar(character.Value, out var byChar))
                {
                    return FromDictionary(byChar);
                }

                return GlyphDictionary.IsAsciiDigit(character.Value)
                    ? Kind.LiningPlain
                    : Kind.Unknown;
            case NameGlyph name:
                return ClassifyName(name.Value);
            default:
                return Kind.Unknown;
        }
    }

    private static Kind ClassifyName(string name)
    {
        var parts = name.Split('.', StringSplitOptions.RemoveEmptyEntries);
        if (parts.Length == 0)
        {
            return Kind.Unknown;
        }

        var suffixes = parts.Skip(1).ToHashSet(StringComparer.Ordinal);
        if (suffixes.Overlaps(["sc", "smcp", "c2sc"]))
        {
            return Kind.SmallCaps;
        }

        var baseName = parts[0];
        var hasPnum = suffixes.Overlaps(["pnum", "lf"]);
        var hasOnum = suffixes.Overlaps(["onum", "tosf"]);
        var hasOsf = suffixes.Contains("osf");
        var hasTf = suffixes.Contains("tf");
        if (GlyphDictionary.IsDigitName(baseName) && (hasPnum || hasOnum || hasOsf || hasTf))
        {
            if (hasOsf || (hasPnum && hasOnum))
            {
                return Kind.OldstyleProportional;
            }

            if (hasOnum)
            {
                return Kind.OldstyleOnum;
            }

            if (hasPnum)
            {
                return Kind.LiningPnum;
            }

            return Kind.LiningPlain;
        }

        if (GlyphDictionary.TryClassByName(name, out var exact))
        {
            return FromDictionary(exact);
        }

        if (GlyphDictionary.TryClassByName(baseName, out var byBase))
        {
            return FromDictionary(byBase);
        }

        return Kind.Unknown;
    }

    private static Kind FromDictionary(string classification) => classification switch
    {
        "capital" => Kind.Capital,
        "lowercase" => Kind.Lowercase,
        "digit" => Kind.LiningPlain,
        "punctuation" => Kind.Punctuation,
        "nonalphabetic" => Kind.NonAlphabetic,
        _ => Kind.Unknown
    };
}
